### DISCUSSION TITLE
wish for open method

### DISCUSSION THREAD

**Discussion by CBusse** [miurahr/py7zr#466] [Ideas]
May I suggest some more advanced interface into a SevenZipFile with an open method like it is provided in the zipfile module? Supporting read and write mode at minimum.

I am desperately looking into a way to use py7zr to generate bigger data sets that should not be kept in a simple data block/array, but need to be streamed into the archive (not from file system, either from another archive, or for example from a database).

For this I'd  like to have some  kind of a file handle into the archive where I could do writes of flexible sizes (just like a normal file handle) and data types.

While there is the writestr method, this method just does not give me the flexibility I would like to have (it is just not sensible to keep a structure that could go into multiple gigabyte sizes in memory all the time to supply as parameter here).

Do you see an easy path to implement this?

**Comment by miurahr** [comment]
Proposal and test implementation is welcome. Because I do not use memory interface by myself, I do not have a preference for it.
Also because of a nature of 7zip format, it is not possible to implement a file-system like interface that zipfile module provide.
7zip is a solid compression format. when you need last several bytes from a  gigabytes of archive, you need to extract all the previous compressed data and then you get a byte.

**Reply by avm19** [reply]
> 7zip is a solid compression format. when you need last several bytes from a gigabytes of archive, you need to extract all the previous compressed data and then you get a byte

@miurahr Does this mean that it is impossible to read a specific file from a 7z archive in a streaming mode?

I see that the latest (1.0.0-rc2) docs show an [example](https://py7zr.readthedocs.io/en/latest/advanced.html#stream-7z-content)  of how to _pass_ a custom file-like object factory to py7zr so that it writes into it except of writing to the default disk filesystem. 

I wonder if it is possible to _get_ a file-like object from py7zr for reading? Again, just like in this Python stdlib ZipFile [example](https://docs.python.org/3/library/zipfile.html#zipfile.ZipFile.open), where a file-like object supports `read(size=-1)` method? In my use case I extract the contents of `archive.7z` on disk, open each resulting file and pass them as streams. I want to avoid extracting files on disk and read directly from the compressed archive, even if this incurs some Python overhead.
